akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala

```diff
@@ -51,18 +52,25 @@ private[http2] final case class ByteHttp2SubStream(
   initialHeaders:        ParsedHeadersFrame,
   data:                  Source[ByteString, Any],
   correlationAttributes: Map[AttributeKey[_], _] = Map.empty
-) extends Http2SubStream {
-  override def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream =
+) extends Http2SubStream[ByteString] {
+  override def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream[ByteString] =
     copy(correlationAttributes = attributes)
 }
 
 private[http2] final case class ChunkedHttp2SubStream(
   initialHeaders:        ParsedHeadersFrame,
   data:                  Source[HttpEntity.ChunkStreamPart, Any],
   correlationAttributes: Map[AttributeKey[_], _]                 = Map.empty
-) extends Http2SubStream {
-  override def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream =
+) extends Http2SubStream[HttpEntity.ChunkStreamPart] {
+  override def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream[HttpEntity.ChunkStreamPart] =
     copy(correlationAttributes = attributes)
+
+  def createResponseEntity(contentLength: Long, contentType: ContentType): RequestEntity = {
+    // Ignore trailing headers when content-length is defined
+    if (contentLength == 0) HttpEntity.Empty
+    else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, data.collect { case Chunk(bytes, _) => bytes })
+    else HttpEntity.Chunked(contentType, data)
```

Comment:
If contentLength is not 0 and not > 0, when will it ever go here, can it be negative to signify undefined?

Reply:
> can it be negative to signify undefined?

Exactly. I guess we should make this an `OptionVal` to make that clearer - but that might be for a different PR.

Reply:
That would require boxing to make it nullable though, so perhaps this is better even if a bit more obscure.